### PR TITLE
Document non-guarantees for Hash

### LINKED ISSUE #74215: Hash of `Discriminant` now produces different results across CPU architectures
Hi, we just noticed that there has been changes to the inner of `Discriminant`, i.e. `DiscriminantKind`, 
that causes a `Hash` of a discriminant to produce different results across CPU architectures.

E.g.

```rust
#[derive(Hash)]
enum SomeEnum {
    One,
    Two,
    Three,
}
```

When hashing this in the current stable one gets writes that look like this:

```
Hash for One:
[0, 0, 0, 0, 0, 0, 0, 0]

Hash for Two
[1, 0, 0, 0, 0, 0, 0, 0]

Hash for Tree
[2, 0, 0, 0, 0, 0, 0, 0]
```

And depending on the architecture in the current nightly one gets for `armv7-unknown-linux-gnueabihf`:

```
Hash for One:
[0, 0, 0, 0]

Hash for Two
[1, 0, 0, 0]

Hash for Tree
[2, 0, 0, 0,]
```

However for ex `x86_64-unknown-linux-gnu` it is still 8 bytes as before.

This is a problem as e.g. CRC hashes are **no longer compatible** between CPU architectures.
Not sure if this was intended, but it seems like a breaking change to me?

In our case we are sending structures between an embedded `thumbv7em-none-eabihf` and an 
`armv7-unknown-linux-gnueabihf` system where CRCs are calculated based on `#[derive(Hash)]` 
for said structures to be sent.

`rustc --version --verbose`:
```
rustc 1.46.0-nightly (5db778aff 2020-07-09)
binary: rustc
commit-hash: 5db778affee7c6600c8e7a177c48282dab3f6292
commit-date: 2020-07-09
host: x86_64-unknown-linux-gnu
release: 1.46.0-nightly
LLVM version: 10.0


rustc 1.44.1 (c7087fe00 2020-06-17)
binary: rustc
commit-hash: c7087fe00d2ba919df1d813c040a5d47e43b0fe7
commit-date: 2020-06-17
host: x86_64-unknown-linux-gnu
release: 1.44.1
LLVM version: 9.0
```

### PR DESCRIPTION
**Pull request by the8472**
Dependence on endianness and type sizes was reported for enum discriminants in #74215 but it is a more general
issue since for example the default implementation of `Hasher::write_usize` uses native endianness.
Additionally the implementations of library types are occasionally changed as their internal fields
change or hashing gets optimized.

## Question

Should this go on the module level documentation instead since it also concerns `Hasher` to some extent and not just `Hash`?

resolves #74215

### CODE REVIEW

File: library/core/src/hash/mod.rs [hunk 1]
```diff
@@ -164,6 +164,19 @@ mod sip;
 /// `0xFF` byte to the `Hasher` so that the values `("ab", "c")` and `("a",
 /// "bc")` hash differently.
 ///
+/// ## Portability
+///
+/// Due to differences in endianness and type sizes data fed by `Hash` to a `Hasher`
```
Comment (pierwill) [inline library/core/src/hash/mod.rs:169]:
```suggestion
/// Due to differences in endianness and type sizes, data fed by `Hash` to a `Hasher`
```

File: library/core/src/hash/mod.rs [hunk 2]
```diff
@@ -164,6 +164,19 @@ mod sip;
...
+/// should not be considered portable across platforms. Additionally the data passed by most
+/// standard library types should not be considered stable between compiler versions.
+///
+/// This means tests shouldn't probe hard-coded hash values or data fed to a `Hasher` and
+/// instead should check consistency with `Eq`.
+///
+/// Serialization formats intended to he portable between platforms or compiler versions should
```
Comment (pierwill) [inline library/core/src/hash/mod.rs:176]:
```suggestion
/// Serialization formats intended to be portable between platforms or compiler versions should
```

**Review by pierwill — COMMENTED** [review]
This is great!

**Comment by nbdd0121** [comment]
A hasher that overrides all provided methods could provide a portable hash though, right? E.g. always use LE, and delegate `write_[ui]size` to `write_[ui]64`.

**Comment by the8472** [comment]
That could cover some cases, but the `Hash` implementation itself could also be platform-dependent because the type is or perhaps turn a `[usize; N]` into `&[u8]` and pass that to the `Hasher`. So both `Hash` and `Hasher` need to be vetted for portability.

**Review by dtolnay — APPROVED** [review]
This looks great to me. Between here and the module docs, this is the place I would think to look first.

Thanks!

**Pull request merged**